docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,298 @@
+background models:
+
+.. code::
+
+    bkg_map = BackgroundModel().map
+    bkg_sampler = MapEventSampler(bkg_map, t_min=, t_max=)
+    events_bkg = bkg_sample.sample()
+
+
+The `MapEventSampler` returns an `astropy.table.Table` object, with the following
+columns: `TIME`, `RA_TRUE`, `DEC_TRUE`, `ENERGY_TRUE`. Later additional info such
+as `DETX` and `DETY` could be added.
+
+
+IRFEventDistributor
+-------------------
+
+We also propose a class to handle correctly the spatially varying IRFs: the
+`IRFEventDistibutor` class. Based on the `PSFMap` and `EDispMap`
+as an input it interpolates the "correct" IRF at the position of a given event and
+applies it. In more detail the class will calculate the psf and the energy dispersion 
+for each event true position and true energy. The IRFs are assumed to be constant
+and not time-dependent. The time dependency will be handled higher level, by a loop
+over good time-intervals (GTIs). The class would be used as following:
+
+.. code::
+
+    events = vstack([events_src_1, events_src_2, ..., events_src_N])
+    event_distributor = IRFEventDistributor(psf_map, edisp_map)
+
+    events = irf_distributor.sample_psf(events)
+    events = irf_distributor.sample_edisp(events)
+
+Alternatively the IRFs could be applied per model component and be handled in 
+the `MapEventSampler`. However in this case the IRFs would not be applied 
+correctly for e.g. very extended sources, because the IRF changes over region 
+of the source. If we find the approach with the `IRFEventDistibutor` to be not 
+performant enough or problematic in memory use, it can be still used as a
+simple fallback solution.
+
+The `IRFEventDistributor` adds the following columns to the event table:
+`RA`, `DEC` and `ENERGY`.
+
+
+MapDatasetEventSampler
+----------------------
+
+To combine the `MapEventSampler` and `IRFEventDistibutor` objects we finally propose 
+to introduce a `MapDatasetEventSampler`, which executes the sampling from all 
+source components, applies the IRFs, samples from the background model and 
+returns a stacked, combined event list.
+
+This is the proposed high level API:
+
+.. code::
+    
+    dataset_sampler = MapDatasetEventSampler(dataset, pointing)
+    events = dataset_sampler.sample()
+
+This requires to add `t_min` and `t_max` information to `MapDataset` or maybe a 
+`MapDataset.gti` attribute, defining the GTI for the given dataset. In addition 
+it also requires to add a lightcurve model to the `SkyModel`. Alternatively the
+information has to added from the outside:
+
+.. code::
+
+    dataset_sampler = MapDatasetEventSampler(dataset, pointing, t_min, t_max)
+    events = dataset_sampler.sample()
+
+
+The `MapEventSampler` finally returns an `EventList` object and adds the following
+columns to the event table: `EVENT_ID` and `MC_ID`. Where `EVENT_ID` is a unique
+number to identify the event, while `MC_ID` is a unique ID (number or string) to
+identify the model component the event was sampled from.
```

Comment:
Suggest to split out the mention of `EVENT_ID` and `MC_ID` into separate paragraphs.

For `EVENT_ID` I guess you would just put `[0 ... n_samples - 1]` at the end? Suggest to say that.

For `MC_ID` it's not clear if / how to fill it. That question of model identifiers is linked a lot to the modeling PIG which has stalled and the work on model serialisation which hasn't started.

Suggest to put this whole mention of `MC_ID` into the outlook section, so that you can finish up this PIG now.

Reply:
Yes, `EVENT_ID` is only a number which is associated to the sampled event. MC_ID can be a number or a string or whatever which identifies the component.